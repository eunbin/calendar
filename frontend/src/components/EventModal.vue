<template>
  <div class="modal"
       :class="{ show: value }">
    <div class="modal__inner">
      <button class="modal__close" @click="close"></button>
      <h2>일정 {{ isAdd ? '등록' : '수정'}}</h2>
      <div class="form">
        <div>
          <label for="title">일정 제목</label>
          <input v-model="model.title" id="title" type="text">
        </div>
        <div>
          <label for="startDate">시작시간</label>
          <date-pick v-model="model.start.dateTime"
                     start-week-on-sunday
                     pick-time
                     :format="datePickerFormat"
                     id="startDate"
                     @input="onChangeStartDateTime"></date-pick>
        </div>
        <div>
          <label for="endDate">종료시간</label>
          <date-pick v-model="model.end.dateTime"
                     start-week-on-sunday
                     pick-time
                     :format="datePickerFormat"
                     id="endDate"
                     @input="onChangeEndDateTime"></date-pick>
        </div><div>
      </div>
      </div>
      <p v-if="errors.length"
         class="validation">
        <b>아래 에러를 확인해주세요.</b>
        <ul>
          <li v-for="(error, index) in errors"
              :key="index">{{ error }}
          </li>
        </ul>
      </p>
      <div class="modal__action">
        <button @click="close">취소</button>
        <button v-show="!isAdd" class="accent" @click="deleteEvent">삭제</button>
        <button class="info" @click="submit">{{ isAdd ? '등록' : '저장'}}</button>
      </div>
    </div>
  </div>
</template>

<script>
import DatePick from 'vue-date-pick'
import 'vue-date-pick/dist/vueDatePick.css'
import moment from 'moment'
import { cloneDeep } from 'lodash-es'
// import { dateFormat } from '@/date'

export default {
  name: 'EventModal',
  components: { DatePick },
  props: {
    value: {
      type: Boolean,
      default: false
    },
    event: {
      type: Object
    }
  },
  data () {
    return {
      model: null,
      errors: []
    }
  },
  computed: {
    isAdd () {
      return !this.event.id
    },
    datePickerFormat () {
      return 'YYYY-MM-DD HH:mm'
    }
  },
  watch: {
    event: {
      immediate: true,
      handler () {
        this.model = cloneDeep(this.event)
        this.errors = []
      }
    }
  },
  created () {
    this.addEventListeners()
  },
  beforeDestroy () {
    this.removeEventListeners()
  },
  methods: {
    onChangeStartDateTime (val) {
      const startDateTime = moment(val)
      const endDateTime = startDateTime.clone().add(1, 'hour')
      this.model.start.date = startDateTime.format('YYYY-MM-DD')
      this.model.end.date = endDateTime.format('YYYY-MM-DD')
      this.model.end.dateTime = endDateTime.format('YYYY-MM-DD HH:mm')
    },
    onChangeEndDateTime (val) {
      const endDateTime = moment(val)
      const startDateTime = endDateTime.clone().subtract(1, 'hour')
      this.model.end.date = endDateTime.format('YYYY-MM-DD')
      this.model.start.date = startDateTime.format('YYYY-MM-DD')
      this.model.start.dateTime = startDateTime.format('YYYY-MM-DD HH:mm')
    },
    submit () {
      if (this.isValidForm()) {
        this.$emit('submit', this.model, this.isAdd)
      }
    },
    deleteEvent () {
      this.$emit('delete')
    },
    close () {
      this.$emit('close')
    },
    isValidForm () {
      if (this.model.title && this.model.start && this.model.end) {
        return true
      }
      this.errors = []
      if (!this.model.title) {
        this.errors.push('일정 제목 입력은 필수입니다.')
      }
      if (!this.model.start) {
        this.errors.push('시작날짜 입력은 필수입니다.')
      }
      if (!this.model.end) {
        this.errors.push('종료날짜 입력은 필수입니다.')
      }
      return false
    },
    handleKeyDown (e) {
      const keyCode = e.keyCode
      if (keyCode === 27) {
        this.close()
      }
    },
    handleClick (e) {
      // FIXME: date-picker close
      if (!e.target.closest('.modal__inner')) {
        this.close()
      }
    },
    addEventListeners () {
      document.addEventListener('keydown', this.handleKeyDown)
      document.addEventListener('click', this.handleClick)
    },
    removeEventListeners () {
      document.addEventListener('keydown', this.handleKeyDown)
      document.addEventListener('click', this.handleClick)
    }
  }
}
</script>

<style lang="scss" scoped>
  .modal {
    opacity: 0;
    visibility: hidden;
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    text-align: left;
    background: rgba(0, 0, 0, .5);
    transition: opacity .25s ease;
  }

  .modal__bg {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    cursor: pointer;
  }

  .modal-state {
    display: none;
  }

  .show  {
    opacity: 1;
    visibility: visible;
  }

  .modal__inner {
    transition: top .25s ease;
    position: absolute;
    top: -20%;
    right: 0;
    bottom: 0;
    left: 0;
    width: 50%;
    margin: auto;
    overflow: auto;
    background: #fff;
    border-radius: 5px;
    padding: 1em 2em;
    height: 700px;

    & .form {
      margin-top: 1em;
      margin-bottom: 1em;
      input {
        width: 100%;
        padding: 12px 20px;
        margin: 8px 0;
        display: inline-block;
        border: 1px solid #ccc;
        border-radius: 4px;
        box-sizing: border-box;
        font-size: 14px;
      }
    }
    .validation {
      color: var(--color-accent);
      font-size: .8em;
    }
    .modal__action {
      display: flex;
      justify-content: flex-end;
      > button {
        border: none;
        background-color: var(--color-gray);
        padding: 16px 32px;
        text-decoration: none;
        margin: 4px 2px;
        cursor: pointer;
        &.info {
          background-color: var(--color-info);
          color: white;
        }
        &.accent {
          background-color: var(--color-accent);
          color: white;
        }
      }
    }
  }

  .modal__close {
    position: absolute;
    right: 1em;
    top: 1em;
    width: 1.1em;
    height: 1.1em;
    cursor: pointer;
  }

  .modal__close:after,
  .modal__close:before {
    content: '';
    position: absolute;
    width: 2px;
    height: 1.5em;
    background: #ccc;
    display: block;
    transform: rotate(45deg);
    left: 50%;
    margin: -3px 0 0 -1px;
    top: 0;
  }

  .modal__close:hover:after,
  .modal__close:hover:before {
    background: #aaa;
  }

  .modal__close:before {
    transform: rotate(-45deg);
  }
</style>
