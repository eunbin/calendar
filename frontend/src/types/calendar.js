const MONTH = 'month'
const WEEK = 'week'

const viewTypes = Object.freeze({
  MONTH,
  WEEK
})

export {
  viewTypes
}
