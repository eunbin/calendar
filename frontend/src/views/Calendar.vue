<template>
  <div class="calendar">
    <calendar-header
      :view-type="viewType"
      :current-date="currentDate"
      @change="onChangeViewType"
      @prev="onPrev"
      @next="onNext"></calendar-header>
    <calendar-view
      :view-type="viewType"
      :current-date="currentDate"
      :selected-date="selectedDate"
      :events="events"
      @event-selected="onSelectEvent"
      @day-selected="onSelectDay"
      @day-changed="onChangeDay"></calendar-view>
    <event-modal v-model="dialog"
                 :event="dialogModel"
                 @submit="onSubmit"
                 @close="onCloseDialog"></event-modal>
  </div>
</template>

<script>
import CalendarHeader from '@/components/CalendarHeader'
import CalendarView from '@/components/CalendarView'
import EventModal from '@/components/EventModal'
import { viewType } from '@/constant/constant'
import moment from 'moment'
import api from '@/service/api'

export default {
  data () {
    return {
      viewType: viewType.month,
      events: [],
      currentDate: null,
      selectedDate: null,
      selectedEvent: null,
      dialog: false,
      dialogModel: {}
    }
  },
  components: {
    EventModal,
    CalendarView,
    CalendarHeader
  },
  computed: {
  },
  watch: {
    currentDate () {
      this.getEvents()
    }
  },
  created () {
    this.selectedDate = this.currentDate

    this.currentDate = moment()
    this.getEvents()
  },
  methods: {
    async getEvents () {
      const year = this.currentDate.year()
      const month = this.currentDate.month() + 1
      try {
        const res = await api.getEvents(year, month)
        this.events = res.data
      } catch (e) {
        console.error(e)
      }
    },
    onSelectEvent (event) {
      this.events.filter(event => event.selected).forEach(event => (event.selected = false))
      this.$set(event, 'selected', !event.selected)
      this.dialog = true
      this.dialogModel = event
    },
    onSelectDay (day) {
      this.dialog = true
    },
    onChangeViewType (viewType) {
      this.viewType = viewType
    },
    async onChangeDay ({ id }, { startDate, endDate }) {
      try {
        await api.updateEventDateById({ id }, { startDate, endDate })
        this.getEvents()
      } catch (e) {
        console.error(e)
      }
    },
    onPrev () {
      if (this.viewType === viewType.month) {
        this.currentDate = this.currentDate.clone().add(-1, 'month')
      }
    },
    onNext () {
      if (this.viewType === viewType.month) {
        this.currentDate = this.currentDate.clone().add(1, 'month')
      }
    },
    async onSubmit (newEvent, isAdd) {
      if (isAdd) {
        await api.addEvent(newEvent)
        this.getEvents()
      } else {
        const { id } = newEvent
        try {
          await api.updateEvent(id, newEvent)
          this.getEvents()
        } catch (e) {
          console.error(e)
        }
      }
    },
    onCloseDialog () {
      this.dialog = false
      this.dialogModel = {}
    }
  }
}
</script>

<style lang="scss" scoped>
  .calendar {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    padding: 1em;
  }
</style>
