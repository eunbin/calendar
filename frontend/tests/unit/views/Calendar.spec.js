import Vuex from 'vuex'
import { createLocalVue, shallowMount } from '@vue/test-utils'
import { DEFAULT_VIEW_TYPE } from '@/types/calendar'
import Calendar from '@/views/Calendar'
import EventModal from '@/components/EventModal'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Calendar.vue', () => {
  let actions
  let store

  beforeEach(() => {
    actions = {
      getEvents: jest.fn()
    }
    store = new Vuex.Store({
      state: {},
      actions
    })
  })

  it('Created Hook 에 getEvents()가 호출되어야 한다.', () => {
    expect(typeof Calendar.created).toBe('function')

    const option = {
      localVue,
      store
    }

    shallowMount(Calendar, option)

    expect(actions.getEvents).toHaveBeenCalled()
  })

  it('기본 보기모드는 DEFAULT_VIEW_TYPE 과 같아야 한다.', () => {
    expect(typeof Calendar.data).toBe('function')

    const defaultData = Calendar.data()

    expect(defaultData.viewType).toBe(DEFAULT_VIEW_TYPE)
  })

  it('페이지에서 열린 에러 모달에서, close 이벤트가 오면 값이 페이지에 업데이트 된다.', () => {
    const option = {
      localVue,
      store
    }

    const wrapper = shallowMount(Calendar, option)
    wrapper.setData({ dialog: true })
    const eventModalWrapper = wrapper.find(EventModal)
    eventModalWrapper.vm.$emit('close', false)

    expect(wrapper.vm.dialog).toBe(false)
  })
})
