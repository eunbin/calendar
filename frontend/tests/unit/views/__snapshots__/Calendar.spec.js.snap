// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar.vue snapshot check 1`] = `
<div
  class="calendar"
>
  <calendar-header-stub
    currentdate="Mon Feb 17 2020 00:00:00 GMT+0900"
    viewtype="month"
  />
   
  <calendar-view-stub
    currentdate="Mon Feb 17 2020 00:00:00 GMT+0900"
    today="Mon Feb 17 2020 00:00:00 GMT+0900"
    viewtype="month"
  />
   
  <event-modal-stub
    event="[object Object]"
  />
</div>
`;
